### DISCUSSION TITLE
مشکل در متد CancelAsync

### DISCUSSION THREAD

**Discussion by raymon-2sam** [Sina-Soltani/Parbad#179] [Q&A]
سلام وقت بخیر
ابتدا تشکر میکنم به خاطر پرباد عالی

یه مشکلی که من جدیدا پیدا کردم این که وقتی متد 
CancelAsync
صدا میزنم عمل نمیکنه و تراکنش کنسل نمیشه
جالب این جاست متد 
VerifyAsync
صدا نمیزنم ولی تو درگاه پرداخت ملت نتیجه تراکنش تایید میشه

پیامی که بعد از صدا زدن متد کنسل بهم میده 
The requested payment is already processed before.

نتیجه cancelResult.IsSucceed هم false

نحوه پیاده سازیشم دقیقا مثل مثال کد خودتونه
لطفا بررسی و راهنمایی بفرمایید

               var cancelResult = await _onlinePayment.CancelAsync(invoice, cancellationReason: "عدم موجودی کالا");
                _helperFile.Log("Cancel the payment ShopId:" + shopDbF.Id + " TrackingNumber:" + invoice.TrackingNumber + " cancelResult" + cancelResult.IsSucceed + "  Message " + cancelResult.Message, "/cancelResult1.txt");

نتیجه متد لاگ به این صورت است
15:04 Cancel the payment ShopId:398 TrackingNumber:371 cancelResultFalse  Message The requested payment is already processed before.

**Comment by Sina-Soltani** [comment]
درود

> جالب این جاست متد
> VerifyAsync
> صدا نمیزنم ولی تو درگاه پرداخت ملت نتیجه تراکنش تایید میشه
> 

آیا مبلغی که توسط خریدار پرداخت شد، به حساب شما واریز شد؟

**Reply by raymon-2sam** [reply]
بله پول به حساب واریز شده و داخل سایت ملت هم 
Successful 
نمایش داده میشه

احتمالش هست چون متد کنسل مشکل داره وقتی که کنسل صدا زده نمیشه به صورت خودکار تایید بشه برای مثلا بعد از 15 دقیقه؟

**Reply by Sina-Soltani** [reply]
لطفا چک کنید آیا مبلغ هنوز داخل حساب شما هست و یا اینکه برگشت داده شده
چون چنین چیزی امکان نداره که بدون استفاده از متد
Verify
مبلغ در حساب فروشنده باقی بمونه

**Reply by raymon-2sam** [reply]
مبلغ که درکارت باقی مونده و برگشت داده نشده
لاگ جدید برای اجرای متد 
Verify
نوشتم که اگر دوباره این مشکل پیش اومد لاگش در دسترسم باشه و بتونم به شما دقیق تر گزارش بدم

الان فقط سوالی که باقی میمونه دلیل عدم کار کردن متد 
CancelAsync

آیا میشه به این دلیل باشه که متد 
Verify
صدا زده شده و بعد از آن دیگه این متد کار نمیکنه؟

سوال دیگه ای که هست این که 
       
       
       `     var invoice = await _onlinePayment.FetchAsync();

            // Check if the invoice is new or it's already processed before.
            if (invoice.Status == PaymentFetchResultStatus.AlreadyProcessed)
            {
      
`
من نمیتونم به متغییر invoice.Status دسترسی داشته باشم و اصلا برام نیست
آیا در ورژن های جدید این ویژگی اضافه شده
ورژن کنونی پرباد بنده 3.2.3 
اگر بله به چه صورت میتونم تو این ورژن همچین if بنویسم

**Reply by Sina-Soltani** [reply]
شما در ابتدا گفتید که متد
Verify
رو اصلا استفاده نکرده بودید
>
> جالب این جاست متد
> VerifyAsync
> صدا نمیزنم ولی تو درگاه پرداخت ملت نتیجه تراکنش تایید میشه

و الان میگید که استفاده کردید
> آیا میشه به این دلیل باشه که متد
> Verify
> صدا زده شده و بعد از آن دیگه این متد کار نمیکنه؟

بالاخره استفاده شده یا نشده؟



همونطور که گفتم، زمانیکه متد
Verify
صدا زده بشه، تراکنش تکمیل میشه و به بانک هم اعلام میشه که تراکنش تکمیل شده
در نتیجه مبلغ در حساب فروشنده باقی میمونه

به هیچ وجه امکان نداره که این متد استفاده نشده باشه و بعدش مبلغ برگشت نخورده باشه
(مدت زمان برگشت برای هر درگاه متفاوت هست. بین پانزده دقیقه تا سه روز)

[نمونه کدهای موجود](https://github.com/Sina-Soltani/Parbad/blob/master/samples/Parbad.Sample.AspNetCore/Controllers/PaymentController.cs#L70)، به صورت کاملا شفاف داره نشون میده که زمانیکه از همه چیز اطمینان دارید، باید این متد رو استفاده کنید
و در صورتیکه قصد دارید تراکنش رو ناموفق ثبت کنید، نباید این متد رو استفاده کنید.
برای نمونه: زمانیکه خریدار پس از پرداخت به سایت شما هدایت شده، موجودی محصولی که سفارش داده بوده در فروشگاه شما تمام شده، پس بنابراین شما متد
Verify
رو نباید استفاده کنید

متد
CancelAsync
پروسه عملیات داخلی پرداخت در سمت شما رو کنسل میکنه و چیزی به درگاه بانک ارسال نمیکنه
به عبارتی، داده های داخلی در مخزن داده رو به روز رسانی میکنه و عملیات رو کنسل شده ثبت میکنه

>
> من نمیتونم به متغییر invoice.Status دسترسی داشته باشم و اصلا برام نیست
> آیا در ورژن های جدید این ویژگی اضافه شده
> ورژن کنونی پرباد بنده 3.2.3
> اگر بله به چه صورت میتونم تو این ورژن همچین if بنویسم

به خاطر ندارم دقیقا در چه نسخه ای اضافه شده
نسخه ای که شما دارید، مال یک سال قبل بوده
سعی کنید همیشه پکیج ها رو آپدیت کنید، نه تنها به خاطر ویژگی هایی که اضافه میشه، بلکه به خاطر باگ هایی که برطرف میشه

**Reply by raymon-2sam** [reply]
جناب سلطانی عزیز من هنوزم میگم این متد فراخوانی نشده طبق کدهایی که نوشته ام ولی خب چون شما مطمن گفتین غیر ممکنه من بی ادبی نکردم وگفنتم لاگی جدیدی براش نوشتم که از این به بعد همه فرخوانی های این متدبرام لاگ بشه و بتونم مطمن تر صحبت کنم ولی طبق کدی که میبینم و دیباگ کردم اصلا این متد فراخوانی نشده

ولی خب شما نوشتین سه روز خب دیشب این اتفاق افتاده و اصلا تا الان اتفاق نیوفتاده بوده ممکن هنوز برگشت بخوره

بابت باقی اطلاعاتتونم ممنونم بله حتما به روزرسانی میشه به خاطر هاست مشتری و نسخه جدید که باید به 
.net core 
میدادم نتونسته بودم ارتقا بدم که الان باهاشون صحبت میکنم هاستشون عوض بکنن و ارتقا انجام بشه

سپاس از شما

**Reply by Sina-Soltani** [reply]
خواهش میکنم
موفق باشید
